File: day13/maze.pl
#!/usr/bin/env perl
use strict;
use warnings;
use 5.24.0;
use feature 'signatures';
no warnings "experimental::signatures";

my ($endx, $endy, $favorite) = @ARGV;

# show_maze(10, 10, $favorite);

my %seen;
my @q = [1,1,0];
my $n = 0;

while (@q) {
    my ($x, $y, $step) = @{shift @q};

    $n++;
    say "$n iterations" if $n % 10 == 0;

    if ($x == $endx && $y == $endy) {
        say "finished in $step steps and $n iterations!";
        last;
    } elsif ($x < 0 || $y <0) {
        next;
    } elsif (is_wall($x, $y, $favorite)) {
        next;
    } elsif (defined $seen{"$x $y"}) {
        next;
    } else {
        $seen{"$x $y"} = 1;
        push @q, [$x-1, $y, $step+1];
        push @q, [$x+1, $y, $step+1];
        push @q, [$x, $y-1, $step+1];
        push @q, [$x, $y+1, $step+1];
    }
}

say "failed";

sub is_wall($x, $y, $favorite) {
    my $val = $x*$x + 3*$x + 2*$x*$y + $y + $y*$y + $favorite;
    my $bin = sprintf '%b', $val;
    my $count = ($bin =~ tr/1//);
    return $count % 2;
}

sub show_maze($width, $height, $favorite) {
    print "  ";
    print $_ % 10 for 0..$width-1;
    print "\n";

    for my $y (0..$height-1) {
        print $y % 10, ' ';
        for my $x (0..$width-1) {
            print is_wall($x, $y, $favorite) ? '#' : '.';
        }
        print "\n";
    }
}

File: day13/maze.cpp
#include <stdio.h>
#include <stdlib.h>
#include <queue>
#include <set>
#include <string>

using namespace std;

struct maze_info {
        int x;
        int y;
        int step;
};

bool is_wall(int x, int y, int favorite);
int num_bits(int n);
string seen_key(int x, int y);

int main(int argc, char *argv[]) {
        queue<struct maze_info> q;
        set<string> seen;
        struct maze_info mi;

        int endx = atoi(argv[1]);
        int endy = atoi(argv[2]);
        int favorite = atoi(argv[3]);

        mi = (struct maze_info) {1, 1, 0};
        q.push(mi);
        int n = 0;
        
        while (!q.empty()) {
                n++;
                mi = q.front();
                q.pop();
                if (n % 10 == 0)
                        printf("%d iterations\n", n);

                if (mi.x == endx && mi.y == endy) {
                        printf("finished in %d steps and %d iterations!\n", mi.step, n);
                        return 0;
                } else if (mi.x < 0 || mi.y < 0) {
                        continue;
                } else if (is_wall(mi.x, mi.y, favorite)) {
                        continue;
                } else {
                        string k = seen_key(mi.x, mi.y);
                        if (seen.find(k) != seen.end()) {
                                continue;
                        } else {
                                seen.insert(k);
                                q.push((struct maze_info) { mi.x-1, mi.y, mi.step+1 });
                                q.push((struct maze_info) { mi.x+1, mi.y, mi.step+1 });
                                q.push((struct maze_info) { mi.x, mi.y-1, mi.step+1 });
                                q.push((struct maze_info) { mi.x, mi.y+1, mi.step+1 });
                        }
                }
        }
        puts("failed");
        return 1;
}

bool is_wall(int x, int y, int favorite) {
        int val = x*x + 3*x + 2*x*y + y + y*y + favorite;
        int count = num_bits(val);
        return count % 2;
}

int num_bits(int n) {
        int count = 0;

        while (n) {
                count += n & 1;
                n >>= 1;
        }

        return count;
}

string seen_key(int x, int y) {
        char s[80];
        sprintf(s, "%d %d", x, y);
        return string(s);
}
